### PR TITLE
feat: `avm/res/compute/gallery`: Added UDT applicationsType

### PR DESCRIPTION
**Pull request by johnlokerse**
## Description

This PR adds the `applicationsType` UDT and updates existing UDTs to follow AVM standards.

Tagging owner: @ReneHezser 

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.compute.gallery](https://github.com/johnlokerse/bicep-registry-modules/actions/workflows/avm.res.compute.gallery.yml)      |

## Type of Change

- [x] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/compute/gallery/application/main.bicep [hunk 1]
```diff
@@ -124,28 +124,34 @@ output location string = application.location
 //   Definitions   //
 // =============== //
 
-type roleAssignmentType = {
-  @sys.description('Optional. The name (as GUID) of the role assignment. If not provided, a GUID will be generated.')
-  name: string?
+import { roleAssignmentType } from 'br/public:avm/utl/types/avm-common-types:0.3.0'
```
Comment (ReneHezser) [inline avm/res/compute/gallery/application/main.bicep:127]: While this is a valid place to import external UDTs, we agreed to import them at the location where they are used. This way it is clear where they come from.
I'd put the import roleAssignmentType right before the param roleAssignments. 
This applies to all imports and references of imported types (AVM common or your own).

File: avm/res/compute/gallery/main.bicep [hunk 2]
```diff
@@ -214,43 +214,9 @@ output imageResourceIds array = [
...
-  @sys.description('Optional. Version of the condition.')
-  conditionVersion: '2.0'?
-
-  @sys.description('Optional. The Resource Id of the delegated managed identity resource.')
-  delegatedManagedIdentityResourceId: string?
-}[]
-
 import { identifierType, purchasePlanType, resourceRangeType } from './image/main.bicep'
+import { customActionType } from './application/main.bicep'
+import { roleAssignmentType, lockType } from 'br/public:avm/utl/types/avm-common-types:0.3.0'
```
Comment (ReneHezser) [inline avm/res/compute/gallery/main.bicep:219]: Please move the imports to the place where they are used and import only one type per line. This way it is possible to use different versions, if needed.

**Review by ReneHezser — COMMENTED** [review]
Thank you for your contribution. I added some small comments.

File: avm/res/compute/gallery/application/main.bicep (hunk 1 above)
Comment (johnlokerse) [reply avm/res/compute/gallery/application/main.bicep:127]: 💡 Thanks for the feedback! Is the single line import required for this type as well?
`import { identifierType, purchasePlanType, resourceRangeType } from './image/main.bicep'`

Comment (johnlokerse) [reply avm/res/compute/gallery/application/main.bicep:127]: Nevermind, these are used within a UDT so must be in the one-liner :)

**Comment by johnlokerse** [comment]
> Thank you for your contribution. I added some small comments.

Thanks! Implemented the comments. Could you re-check it to see if it's as expected?

**Review by ReneHezser — DISMISSED** [review]
Thanks again for your contribution.

For next time, please do not solve conversations. Usually, the reviewer does that after checking again. This makes it easier for us to do the reviews on many modules at the same time.

File: avm/res/compute/gallery/application/main.bicep (hunk 1 above)
Comment (AlexanderSehr) [reply avm/res/compute/gallery/application/main.bicep:127]: One point may have gotten lost a bit. 

Hey @johnlokerse, long time no see. 😄 To build on top of what @ReneHezser said: Just for the types that you import from other templates, please import them on top of where they are used. For example:

```bicep
import { lockType } from 'br/public:avm/utl/types/avm-common-types:0.3.0'
@sys.description('Optional. The lock settings of the service.')
param lock lockType?
```

Comment (AlexanderSehr) [reply avm/res/compute/gallery/application/main.bicep:127]: Just makes it a bit easier to find - especially if the module would happen to not have any other types :) (in know that's not the case for you, but I wanted to mention it)

Comment (johnlokerse) [reply avm/res/compute/gallery/application/main.bicep:127]: Done! Also, I have moved the imports at the UDTs to above the `@export()`:

CleanShot 2024-11-25 at 09 18 18@2x

Comment (AlexanderSehr) [reply avm/res/compute/gallery/application/main.bicep:127]: Great, thanks @johnlokerse. 🚀 
Just one ask - please check if you pushed your latest changes. In the PR I can see that e.g. the 'roleAssignmentType' is not on top of the parameter (yet others are).

File: avm/res/compute/gallery/main.bicep [hunk 3]
```diff
@@ -316,3 +281,38 @@ type imageType = {
...
+  privacyStatementUri: string?
+
+  @sys.description('Optional. The release note uri. Has to be a valid URL.')
+  releaseNoteUri: string?
+
+  @sys.description('Optional. The end of life date of the gallery application definition. This property can be used for decommissioning purposes. This property is updatable.')
+  endOfLifeDate: string?
+
+  @sys.description('Optional. Array of role assignments to create.')
+  roleAssignments: roleAssignmentType[]?
```
Comment (johnlokerse) [inline avm/res/compute/gallery/main.bicep:311]: @AlexanderSehr Do you mean here? If yes, do you want to see the import above the UDT definition or above the property inside the UDT? Also, the `roleAssignmentType` will be imported twice.

Comment (johnlokerse) [reply avm/res/compute/gallery/main.bicep:311]: Import definition inside the UDT is not possible. It throws an error.

Comment (AlexanderSehr) [reply avm/res/compute/gallery/main.bicep:311]: Hey @johnlokerse, not quite. In fact in the template I commented in. Will attach a screenshot :)

**Comment by AlexanderSehr** [comment]
Hey @ReneHezser, I'll leave the final approval to you as you reviewed the PR initially :)

**Comment by johnlokerse** [comment]
@ReneHezser Let me know if I have to do something 😃

**Review by ReneHezser — APPROVED** [review]
Thanks again for the contribution.

**Comment by ReneHezser** [comment]
@johnlokerse You can merge the PR now. Thanks for your patience.

**Comment by johnlokerse** [comment]
@AlexanderSehr or @ReneHezser Can you merge it for me? I am just a contributor ;-)

**Comment by AlexanderSehr** [comment]
Say no more @johnlokerse

**Pull request merged**